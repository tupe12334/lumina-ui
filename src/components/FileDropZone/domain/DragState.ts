export type DragStateType = 'idle' | 'dragOver' | 'dragLeave'

export class DragState {
  private constructor(private readonly state: DragStateType) {}

  public static create(state: DragStateType = 'idle'): DragState {
    return new DragState(state)
  }

  public getState(): DragStateType {
    return this.state
  }

  public isIdle(): boolean {
    return this.state === 'idle'
  }

  public isDragOver(): boolean {
    return this.state === 'dragOver'
  }

  public isDragLeave(): boolean {
    return this.state === 'dragLeave'
  }

  public isActive(): boolean {
    return this.state === 'dragOver'
  }

  public setDragOver(): DragState {
    return new DragState('dragOver')
  }

  public setDragLeave(): DragState {
    return new DragState('dragLeave')
  }

  public setIdle(): DragState {
    return new DragState('idle')
  }
}