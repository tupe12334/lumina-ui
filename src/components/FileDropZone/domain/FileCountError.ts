import { FileValidationError } from './FileValidationError'

export class FileCountError extends FileValidationError {
  constructor(actualCount: number, maxCount: number) {
    super(`Cannot upload ${actualCount} files. Maximum allowed: ${maxCount}`)
    this.name = 'FileCountError'
  }
}