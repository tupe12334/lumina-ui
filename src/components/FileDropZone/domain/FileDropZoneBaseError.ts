export class FileDropZoneError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly fileId?: string
  ) {
    super(message)
    this.name = 'FileDropZoneError'
  }
}