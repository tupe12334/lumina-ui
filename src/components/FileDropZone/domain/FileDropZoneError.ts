export class FileDropZoneError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly fileId?: string
  ) {
    super(message)
    this.name = 'FileDropZoneError'
  }
}

export class FileValidationError extends FileDropZoneError {
  constructor(message: string, fileId?: string) {
    super(message, 'VALIDATION_ERROR', fileId)
    this.name = 'FileValidationError'
  }
}

export class FileUploadError extends FileDropZoneError {
  constructor(message: string, fileId?: string) {
    super(message, 'UPLOAD_ERROR', fileId)
    this.name = 'FileUploadError'
  }
}

export class FileSizeError extends FileValidationError {
  constructor(actualSize: number, maxSize: number, fileId?: string) {
    const actualSizeMB = Math.round(actualSize / 1024 / 1024)
    const maxSizeMB = Math.round(maxSize / 1024 / 1024)
    super(`File size ${actualSizeMB}MB exceeds maximum allowed size of ${maxSizeMB}MB`, fileId)
    this.name = 'FileSizeError'
  }
}

export class FileTypeError extends FileValidationError {
  constructor(actualType: string, allowedTypes: string[], fileId?: string) {
    const typesList = allowedTypes.join(', ')
    super(`File type "${actualType}" is not allowed. Allowed types: ${typesList}`, fileId)
    this.name = 'FileTypeError'
  }
}

export class FileCountError extends FileValidationError {
  constructor(actualCount: number, maxCount: number) {
    super(`Cannot upload ${actualCount} files. Maximum allowed: ${maxCount}`)
    this.name = 'FileCountError'
  }
}