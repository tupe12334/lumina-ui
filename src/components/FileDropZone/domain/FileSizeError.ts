import { FileValidationError } from './FileValidationError'

export class FileSizeError extends FileValidationError {
  constructor(actualSize: number, maxSize: number, fileId?: string) {
    const actualSizeMB = Math.round(actualSize / 1024 / 1024)
    const maxSizeMB = Math.round(maxSize / 1024 / 1024)
    super(`File size ${actualSizeMB}MB exceeds maximum allowed size of ${maxSizeMB}MB`, fileId)
    this.name = 'FileSizeError'
  }
}