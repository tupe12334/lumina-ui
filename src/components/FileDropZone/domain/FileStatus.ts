import { type FileStatus } from './FileStatusType'

export class FileStatusValue {
  private constructor(private readonly status: FileStatus) {}

  public static create(status?: FileStatus): FileStatusValue {
    const defaultStatus = status || 'pending'
    return new FileStatusValue(defaultStatus)
  }

  public getStatus(): FileStatus {
    return this.status
  }

  public isPending(): boolean {
    return this.status === 'pending'
  }

  public isUploading(): boolean {
    return this.status === 'uploading'
  }

  public isUploaded(): boolean {
    return this.status === 'uploaded'
  }

  public isProcessing(): boolean {
    return this.status === 'processing'
  }

  public isCompleted(): boolean {
    return this.status === 'completed'
  }

  public isFailed(): boolean {
    return this.status === 'failed'
  }

  public isActive(): boolean {
    return this.status === 'uploading' || this.status === 'processing'
  }

  public canRetry(): boolean {
    return this.status === 'failed'
  }
}