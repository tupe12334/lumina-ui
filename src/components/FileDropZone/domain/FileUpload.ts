import { FileStatusValue, type FileStatus } from './FileStatus'

export interface FileUploadData {
  id: string
  file: File
  status: FileStatus
  progress: number
  error?: string
  previewUrl?: string
}

export class FileUpload {
  private constructor(
    private readonly id: string,
    private readonly file: File,
    private readonly status: FileStatusValue,
    private readonly progress: number,
    private readonly error?: string,
    private readonly previewUrl?: string
  ) {}

  public static create(file: File, id?: string): FileUpload {
    const fileId = id || `${file.name}-${file.size}-${Date.now()}`
    const status = FileStatusValue.create('pending')
    return new FileUpload(fileId, file, status, 0)
  }

  public static fromData(data: FileUploadData): FileUpload {
    const status = FileStatusValue.create(data.status)
    return new FileUpload(
      data.id,
      data.file,
      status,
      data.progress,
      data.error,
      data.previewUrl
    )
  }

  public getId(): string {
    return this.id
  }

  public getFile(): File {
    return this.file
  }

  public getStatus(): FileStatusValue {
    return this.status
  }

  public getProgress(): number {
    return this.progress
  }

  public getError(): string | undefined {
    return this.error
  }

  public getPreviewUrl(): string | undefined {
    return this.previewUrl
  }

  public getName(): string {
    return this.file.name
  }

  public getSize(): number {
    return this.file.size
  }

  public getType(): string {
    return this.file.type
  }

  public withStatus(status: FileStatus): FileUpload {
    const newStatus = FileStatusValue.create(status)
    return new FileUpload(
      this.id,
      this.file,
      newStatus,
      this.progress,
      this.error,
      this.previewUrl
    )
  }

  public withProgress(progress: number): FileUpload {
    const clampedProgress = Math.max(0, Math.min(100, progress))
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      clampedProgress,
      this.error,
      this.previewUrl
    )
  }

  public withError(error: string): FileUpload {
    const failedStatus = FileStatusValue.create('failed')
    return new FileUpload(
      this.id,
      this.file,
      failedStatus,
      this.progress,
      error,
      this.previewUrl
    )
  }

  public withPreviewUrl(previewUrl: string): FileUpload {
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      this.progress,
      this.error,
      previewUrl
    )
  }

  public clearError(): FileUpload {
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      this.progress,
      undefined,
      this.previewUrl
    )
  }

  public toData(): FileUploadData {
    return {
      id: this.id,
      file: this.file,
      status: this.status.getStatus(),
      progress: this.progress,
      error: this.error,
      previewUrl: this.previewUrl
    }
  }
}