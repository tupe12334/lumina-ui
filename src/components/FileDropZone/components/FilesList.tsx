import React from 'react'
import { FileUpload } from '../domain/FileUpload'
import { FileListItem } from './FileListItem'
import styles from '../FileDropZone.module.css'

interface FilesListProps {
  files: FileUpload[]
  onRemoveFile: (fileId: string) => void
  onRetryFile: (fileId: string) => void
  onClearAll: () => void
  showClearAll?: boolean
}

export const FilesList: React.FC<FilesListProps> = ({
  files,
  onRemoveFile,
  onRetryFile,
  onClearAll,
  showClearAll
}) => {
  const shouldShowClearAll = showClearAll !== false
  if (files.length === 0) {
    return null
  }

  return (
    <div className={styles.filesList}>
      <div className={styles.filesHeader}>
        <h3 className={styles.filesTitle}>
          {files.length} {files.length === 1 ? 'file' : 'files'}
        </h3>
        {shouldShowClearAll && (
          <button
            type="button"
            className={styles.clearButton}
            onClick={onClearAll}
            aria-label="Clear all files"
          >
            Clear all
          </button>
        )}
      </div>

      <div className={styles.filesList} role="list" aria-label="Uploaded files">
        {files.map((fileUpload) => (
          <FileListItem
            key={fileUpload.getId()}
            fileUpload={fileUpload}
            onRemoveFile={onRemoveFile}
            onRetryFile={onRetryFile}
          />
        ))}
      </div>
    </div>
  )
}