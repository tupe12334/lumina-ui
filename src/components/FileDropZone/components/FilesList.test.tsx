import { render, screen } from '@testing-library/react'
import { vi } from 'vitest'
import { FilesList } from './FilesList'
import { FileUpload } from '../domain/FileUpload'

const mockFile1 = new File(['test content 1'], 'test1.txt', { type: 'text/plain' })
const mockFile2 = new File(['test content 2'], 'test2.txt', { type: 'text/plain' })
const mockFiles = [
  FileUpload.create(mockFile1, 'test-id-1'),
  FileUpload.create(mockFile2, 'test-id-2'),
]

const defaultProps = {
  files: mockFiles,
  onRemoveFile: vi.fn(),
  onRetryFile: vi.fn(),
  onClearAll: vi.fn(),
}

describe('FilesList', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  test('renders correctly with files', () => {
    const { container } = render(<FilesList {...defaultProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('returns null when no files provided', () => {
    const { container } = render(
      <FilesList
        {...defaultProps}
        files={[]}
      />
    )
    expect(container.firstChild).toBeNull()
  })

  test('displays correct file count', () => {
    render(<FilesList {...defaultProps} />)

    expect(screen.getByText('2 files')).toBeInTheDocument()
  })

  test('displays singular file text for one file', () => {
    render(
      <FilesList
        {...defaultProps}
        files={[mockFiles[0]]}
      />
    )

    expect(screen.getByText('1 file')).toBeInTheDocument()
  })

  test('shows clear all button by default', () => {
    render(<FilesList {...defaultProps} />)

    const clearButton = screen.getByText('Clear all')
    expect(clearButton).toBeInTheDocument()
  })

  test('hides clear all button when showClearAll is false', () => {
    render(
      <FilesList
        {...defaultProps}
        showClearAll={false}
      />
    )

    expect(screen.queryByText('Clear all')).not.toBeInTheDocument()
  })

  test('calls onClearAll when clear button is clicked', () => {
    render(<FilesList {...defaultProps} />)

    const clearButton = screen.getByText('Clear all')
    clearButton.click()

    expect(defaultProps.onClearAll).toHaveBeenCalled()
  })

  test('renders file list with correct accessibility attributes', () => {
    render(<FilesList {...defaultProps} />)

    const fileList = screen.getByRole('list', { name: 'Uploaded files' })
    expect(fileList).toBeInTheDocument()

    const fileItems = screen.getAllByRole('listitem')
    expect(fileItems).toHaveLength(2)
  })

  test('passes correct props to FileListItem components', () => {
    render(<FilesList {...defaultProps} />)

    // Check that both files are rendered
    expect(screen.getByText('test1.txt')).toBeInTheDocument()
    expect(screen.getByText('test2.txt')).toBeInTheDocument()

    // Check that remove buttons are present for both files
    expect(screen.getByLabelText('Remove test1.txt')).toBeInTheDocument()
    expect(screen.getByLabelText('Remove test2.txt')).toBeInTheDocument()
  })

  test('has correct accessibility structure', () => {
    render(<FilesList {...defaultProps} />)

    const filesHeader = screen.getByText('2 files').closest('div')
    expect(filesHeader).toBeInTheDocument()

    const clearButton = screen.getByLabelText('Clear all files')
    expect(clearButton).toBeInTheDocument()
  })
})