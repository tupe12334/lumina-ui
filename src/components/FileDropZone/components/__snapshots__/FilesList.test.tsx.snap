// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilesList > renders correctly with files 1`] = `
<div
  class="_filesList_3af3de"
>
  <div
    class="_filesHeader_3af3de"
  >
    <h3
      class="_filesTitle_3af3de"
    >
      2
       
      files
    </h3>
    <button
      aria-label="Clear all files"
      class="_clearButton_3af3de"
      type="button"
    >
      Clear all
    </button>
  </div>
  <div
    aria-label="Uploaded files"
    class="_filesList_3af3de"
    role="list"
  >
    <div
      class="_fileItem_3af3de"
      role="listitem"
    >
      <div
        class="_filePreview_3af3de"
      >
        <svg
          class="_fileIcon_3af3de"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            class="_fileIcon_3af3de"
            d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
          />
          <polyline
            class="_fileIcon_3af3de"
            points="14,2 14,8 20,8"
          />
        </svg>
      </div>
      <div
        class="_fileInfo_3af3de"
      >
        <p
          class="_fileName_3af3de"
          title="test1.txt"
        >
          test1.txt
        </p>
        <div
          class="_fileDetails_3af3de"
        >
          <span
            class="_fileSize_3af3de"
          >
            14 B
          </span>
          <div
            aria-label="File status: pending"
            class="_fileStatus_3af3de _pending_3af3de"
            role="status"
          >
            pending
          </div>
        </div>
      </div>
      <div
        class="_fileActions_3af3de"
      >
        <button
          aria-label="Remove test1.txt"
          class="_actionButton_3af3de _removeButton_3af3de"
          title="Remove file"
          type="button"
        >
          <svg
            class=""
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
          >
            <polyline
              class=""
              points="3,6 5,6 21,6"
            />
            <path
              class=""
              d="m19,6v14a2,2 0 0,1 -2,2H7a2,2 0 0,1 -2,-2V6m3,0V4a2,2 0 0,1 2,-2h4a2,2 0 0,1 2,2v2"
            />
            <line
              class=""
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              class=""
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="_fileItem_3af3de"
      role="listitem"
    >
      <div
        class="_filePreview_3af3de"
      >
        <svg
          class="_fileIcon_3af3de"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            class="_fileIcon_3af3de"
            d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
          />
          <polyline
            class="_fileIcon_3af3de"
            points="14,2 14,8 20,8"
          />
        </svg>
      </div>
      <div
        class="_fileInfo_3af3de"
      >
        <p
          class="_fileName_3af3de"
          title="test2.txt"
        >
          test2.txt
        </p>
        <div
          class="_fileDetails_3af3de"
        >
          <span
            class="_fileSize_3af3de"
          >
            14 B
          </span>
          <div
            aria-label="File status: pending"
            class="_fileStatus_3af3de _pending_3af3de"
            role="status"
          >
            pending
          </div>
        </div>
      </div>
      <div
        class="_fileActions_3af3de"
      >
        <button
          aria-label="Remove test2.txt"
          class="_actionButton_3af3de _removeButton_3af3de"
          title="Remove file"
          type="button"
        >
          <svg
            class=""
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
          >
            <polyline
              class=""
              points="3,6 5,6 21,6"
            />
            <path
              class=""
              d="m19,6v14a2,2 0 0,1 -2,2H7a2,2 0 0,1 -2,-2V6m3,0V4a2,2 0 0,1 2,-2h4a2,2 0 0,1 2,2v2"
            />
            <line
              class=""
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              class=""
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
